src/components/CodeOverview/index.tsx

```diff
@@ -0,0 +1,217 @@
+    // After a short delay, get the new height and re-render.
+    _window.addEventListener('resize', this.waitAndSetNewOverviewHeight);
+  }
+
+  componentWillUnmount() {
+    const { _window } = this.props;
+    _window.removeEventListener('resize', this.resetOverviewHeight);
+    _window.removeEventListener('resize', this.waitAndSetNewOverviewHeight);
+  }
+
+  resetOverviewHeight = () => {
+    this.setState({ overviewHeight: null });
+  };
+
+  setOverviewHeight = (ref = this.overviewRef) => {
+    if (ref && ref.current) {
+      this.setState({ overviewHeight: ref.current.clientHeight });
+    }
+  };
+
+  waitAndSetNewOverviewHeight = this.props._debounce(
+    () => this.setOverviewHeight(),
+    200,
+    {
+      leading: false,
+      trailing: true,
+    },
+  );
+
+  fitLineShapesIntoOverview(allLineShapes: AllLineShapes) {
+    const { overviewPadding, rowHeight, rowTopPadding } = this.props;
+    const { overviewHeight } = this.state;
+
+    if (!overviewHeight) {
+      throw new Error(
+        'overviewHeight must be set when calling fitLineShapesIntoOverview()',
+      );
+    }
+
+    const availableHeight =
+      overviewHeight -
+      // Remove the top and bottom div padding.
+      overviewPadding * 2 -
+      // Adjust for the first row not having padding.
+      rowTopPadding;
+
+    const numberOfRows = Math.floor(availableHeight / rowHeight);
+
+    let chunkSize = 1;
+    if (allLineShapes.length > numberOfRows) {
+      // Split file lines evenly between all rows.
+      chunkSize = Math.ceil(allLineShapes.length / numberOfRows);
+    }
+
+    return {
+      numberOfRows,
+      chunkedLineShapes: chunk(allLineShapes, chunkSize),
+    };
+  }
+
+  renderRow(
+    selectedMessageMap: LinterProviderInfo['selectedMessageMap'],
+    rowIndex: number,
+    shapeIndex: number,
+    groupOflineShapes: LineShapes[] | void,
+  ) {
+    if (!groupOflineShapes) {
+      return null;
+    }
+
+    // TODO: render linter messages here by looking at all shapes
+    // in groupOflineShapes.
+    // See https://github.com/mozilla/addons-code-manager/issues/523
+
+    const lineShapes = groupOflineShapes[shapeIndex];
+
+    return <CodeLineShapes lineShapes={lineShapes} />;
+  }
+
+  renderOverview(selectedMessageMap: LinterProviderInfo['selectedMessageMap']) {
+    const { content, location, rowHeight, rowTopPadding } = this.props;
+    const { overviewHeight } = this.state;
+
+    if (!overviewHeight) {
+      return null;
+    }
+
+    const allLineShapes = generateLineShapes(getLines(content));
+    const { numberOfRows, chunkedLineShapes } = this.fitLineShapesIntoOverview(
+      allLineShapes,
+    );
+    const overview = [];
+
+    for (let rowIndex = 0; rowIndex < numberOfRows; rowIndex++) {
+      const shapes = chunkedLineShapes[rowIndex] || undefined;
+      // Use the first line in the group.
+      const shapeIndex = 0;
+      const line =
+        shapes && shapes.length ? shapes[shapeIndex].line : undefined;
```

Comment:
How can we end up with an empty `shapes` array?

Reply:
I added a comment in the code to explain it.